/*
------------------------------------------------------------------------------ 
This code was generated by Amplication. 
 
Changes to this file will be lost if the code is regenerated. 

There are other ways to to customize your code, see this doc to learn more
https://docs.amplication.com/how-to/custom-code

------------------------------------------------------------------------------
  */
import * as common from "@nestjs/common";
import * as swagger from "@nestjs/swagger";
import { isRecordNotFoundError } from "../../prisma.util";
import * as errors from "../../errors";
import { Request } from "express";
import { plainToClass } from "class-transformer";
import { ApiNestedQuery } from "../../decorators/api-nested-query.decorator";
import { ReviewRiskFactorOtherService } from "../reviewRiskFactorOther.service";
import { ReviewRiskFactorOtherCreateInput } from "./ReviewRiskFactorOtherCreateInput";
import { ReviewRiskFactorOtherWhereInput } from "./ReviewRiskFactorOtherWhereInput";
import { ReviewRiskFactorOtherWhereUniqueInput } from "./ReviewRiskFactorOtherWhereUniqueInput";
import { ReviewRiskFactorOtherFindManyArgs } from "./ReviewRiskFactorOtherFindManyArgs";
import { ReviewRiskFactorOtherUpdateInput } from "./ReviewRiskFactorOtherUpdateInput";
import { ReviewRiskFactorOther } from "./ReviewRiskFactorOther";

export class ReviewRiskFactorOtherControllerBase {
  constructor(protected readonly service: ReviewRiskFactorOtherService) {}
  @common.Post()
  @swagger.ApiCreatedResponse({ type: ReviewRiskFactorOther })
  async create(
    @common.Body() data: ReviewRiskFactorOtherCreateInput
  ): Promise<ReviewRiskFactorOther> {
    return await this.service.create({
      data: data,
      select: {
        id: true,
        riskFactorOther: true,
      },
    });
  }

  @common.Get()
  @swagger.ApiOkResponse({ type: [ReviewRiskFactorOther] })
  @ApiNestedQuery(ReviewRiskFactorOtherFindManyArgs)
  async findMany(
    @common.Req() request: Request
  ): Promise<ReviewRiskFactorOther[]> {
    const args = plainToClass(ReviewRiskFactorOtherFindManyArgs, request.query);
    return this.service.findMany({
      ...args,
      select: {
        id: true,
        riskFactorOther: true,
      },
    });
  }

  @common.Get("/:id")
  @swagger.ApiOkResponse({ type: ReviewRiskFactorOther })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async findOne(
    @common.Param() params: ReviewRiskFactorOtherWhereUniqueInput
  ): Promise<ReviewRiskFactorOther | null> {
    const result = await this.service.findOne({
      where: params,
      select: {
        id: true,
        riskFactorOther: true,
      },
    });
    if (result === null) {
      throw new errors.NotFoundException(
        `No resource was found for ${JSON.stringify(params)}`
      );
    }
    return result;
  }

  @common.Patch("/:id")
  @swagger.ApiOkResponse({ type: ReviewRiskFactorOther })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async update(
    @common.Param() params: ReviewRiskFactorOtherWhereUniqueInput,
    @common.Body() data: ReviewRiskFactorOtherUpdateInput
  ): Promise<ReviewRiskFactorOther | null> {
    try {
      return await this.service.update({
        where: params,
        data: data,
        select: {
          id: true,
          riskFactorOther: true,
        },
      });
    } catch (error) {
      if (isRecordNotFoundError(error)) {
        throw new errors.NotFoundException(
          `No resource was found for ${JSON.stringify(params)}`
        );
      }
      throw error;
    }
  }

  @common.Delete("/:id")
  @swagger.ApiOkResponse({ type: ReviewRiskFactorOther })
  @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
  async delete(
    @common.Param() params: ReviewRiskFactorOtherWhereUniqueInput
  ): Promise<ReviewRiskFactorOther | null> {
    try {
      return await this.service.delete({
        where: params,
        select: {
          id: true,
          riskFactorOther: true,
        },
      });
    } catch (error) {
      if (isRecordNotFoundError(error)) {
        throw new errors.NotFoundException(
          `No resource was found for ${JSON.stringify(params)}`
        );
      }
      throw error;
    }
  }
}
